tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,748 @@
+)
+
+// NewLiveChunkDiskMapper returns a new writer against the given directory
+// using the default segment time.
+func NewLiveChunkDiskMapper(dir string, pool chunkenc.Pool) (*LiveChunkDiskMapper, error) {
+	return newLiveChunkDiskMapper(dir, DefaultHeadChunkSegmentMaxTimeRange, pool)
+}
+
+func newLiveChunkDiskMapper(dir string, segmentTime int64, pool chunkenc.Pool) (*LiveChunkDiskMapper, error) {
+	if segmentTime <= 0 {
+		segmentTime = DefaultHeadChunkSegmentMaxTimeRange
+	}
+
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return nil, err
+	}
+	dirFile, err := fileutil.OpenDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	hrw := &LiveChunkDiskMapper{
+		dirFile:     dirFile,
+		segmentTime: segmentTime,
+		pool:        pool,
+		crc32:       newCRC32(),
+	}
+
+	for i := 0; i < inBufferShards; i++ {
+		hrw.inBufferChunks[i] = make(map[uint64]chunkenc.Chunk)
+	}
+
+	if err := hrw.openMMapFiles(); err != nil {
+		return nil, err
+	}
+
+	return hrw, nil
+}
+
+func (w *LiveChunkDiskMapper) openMMapFiles() (err error) {
+	bs := map[int]ByteSlice{}
+	cs := map[int]io.Closer{}
+	defer func() {
+		if err != nil {
+			var merr tsdb_errors.MultiError
+			merr.Add(err)
+			merr.Add(closeAllFromMap(cs))
+			err = merr
+		}
+	}()
+
+	files, err := sequenceFilesMap(w.dirFile.Name())
+	if err != nil {
+		return err
+	}
+	if w.pool == nil {
+		w.pool = chunkenc.NewPool()
+	}
+
+	seqs := make([]int, 0, len(files))
+	for seq, fn := range files {
+		f, err := fileutil.OpenMmapFile(fn)
+		if err != nil {
+			return errors.Wrap(err, "mmap files")
+		}
+		cs[seq] = f
+		bs[seq] = realByteSlice(f.Bytes())
+		seqs = append(seqs, seq)
+	}
+
+	w.mmappedChunks = bs
+	w.closers = cs
+	w.size = 0
+
+	// Check for unsequential files.
+	sort.Ints(seqs)
+	if len(seqs) == 0 {
+		return nil
+	}
+	lastSeq := seqs[0]
+	for _, seq := range seqs[1:] {
+		if seq != lastSeq+1 {
+			return errors.Errorf("found unsequential segment files %d and %d", lastSeq, seq)
+		}
+		lastSeq = seq
+	}
+
+	for i, b := range w.mmappedChunks {
+		if b.Len() < HeadSegmentHeaderSize {
+			return errors.Wrapf(errInvalidSize, "invalid segment header in segment %d", i)
+		}
+		// Verify magic number.
+		if m := binary.BigEndian.Uint32(b.Range(0, MagicChunksSize)); m != MagicHeadChunks {
+			return errors.Errorf("invalid magic number %x", m)
+		}
+
+		// Verify chunk format version.
+		if v := int(b.Range(MagicChunksSize, MagicChunksSize+ChunksFormatVersionSize)[0]); v != chunksFormatV1 {
+			return errors.Errorf("invalid chunk format version %d", v)
+		}
+
+		maxt := binary.BigEndian.Uint64(b.Range(HeaderMaxtOffset, HeaderMaxtOffset+8))
+		if maxt == 0 {
+			// This is possible if Prometheus crashes and the maxt was unwritten to the
+			// last segment. As a safe buffer, we set the maxt to mint + 1.5 times segment time range.
+			f, err := os.OpenFile(files[i], os.O_WRONLY|os.O_CREATE, 0666)
+			if err != nil {
+				return err
+			}
+			defer f.Close()
```

Comment:
It would be nice to catch this error with something like https://github.com/thanos-io/thanos/blob/master/pkg/runutil/runutil.go#L133